Hmac.js

```diff
@@ -48,12 +52,24 @@ export class Hmac {
    *
    * @param {object} options - The options to use.
    * @param {Uint8Array} options.data - The data to sign as a Uint8Array.
+   * @param {boolean} options.useCache - Enable the use of a cache.
    *
    * @returns {Promise<string>} The base64url-encoded signature.
    */
-  async sign({data}) {
+  async sign({data, useCache = true}) {
+    if(useCache) {
+      const sig = this.cache.get(base64url.encode(data));
+      if(sig) {
+        return sig;
+      }
+    }
     const {id: keyId, kmsClient, capability, invocationSigner} = this;
-    return kmsClient.sign({keyId, data, capability, invocationSigner});
+    const sig = await kmsClient.sign(
+      {keyId, data, capability, invocationSigner});
+    if(useCache) {
+      this.cache.set(base64url.encode(data), sig);
+    }
+    return sig;
```

Comment:
There's a race and/or inefficiency with this cache pattern.  Multiple concurrent callers could have a cache miss, all do the async work, then all cache the same (I assume) result.  A better pattern would be to cache promises.  Check the cache, if a hit then return the result promise, if a miss then create a new promise, set it in the cache, and return it.  The returned promise would do the async work and resolve to the result.  That ensures only one result calculated per checked key.  I'm not sure what the error flow needs to be here.

Reply:
+1 to using caching a promise here.

> I'm not sure what the error flow needs to be here.

Error flow should be to just remove the entry from the cache and then throw whatever the error was.

Reply:
I intentionally did not implement a request queue. That can be done if it is now being requested.

Reply:
Yeah, let's use promises instead. I would expect it to be a very minor change.